File: python-project-euler-100/p011.py
import time
from Solution import Solution

class P011(Solution):
    def read_grid(self, file_name):
        grid = []
        with open(file_name) as file_in:
            for line in file_in:
                parts = line.split()
                row = []
                for part in parts:
                    row.append(int(part))
                grid.append(row)
        return grid
    
    def grid_product(self, grid, i, j, count):
        n = len(grid)
        max_product = 1
        product = 1
        if count <= (n-i):
            for k in range(i, i + count):
                product *= grid[j][k]
            if product > max_product:
                max_product = product
        
        product = 1
        if count <= (n-j):
            for k in range(j, j + count):
                product *= grid[k][i]
            if product > max_product:
                max_product = product
        
        product = 1
        if (count <= (n-i)) and (count <= (n-j)):
            for k in range(0, count):
                product *= grid[j + k][i + k]
            if product > max_product:
                max_product = product
        
        product = 1
        if (count <= (n-i)) and (j > (count -1)):
            for k in range(0, count):
                product *= grid[j - k][i + k]
            if product > max_product:
                max_product = product
        
        return max_product
    
    def solve(self):
        self.problem_number = 11
        max_product = 1
        count = 4
    
        grid = self.read_grid("p011.txt")
        n = len(grid)
        for i in range(0, n):
            for j in range(0, n):
                product = self.grid_product(grid, i, j, count)
                if product > max_product:
                    max_product = product
    
        return max_product

def main():
	P011().run()

if __name__ == "__main__":
	main()

File: typescript-project-euler-100/src/p011.ts
import { Solution } from "./Solution";
import { readFileSync } from 'fs';

export class P011 extends Solution{

    readGrid(fileName: string): number[][]{
		const file = readFileSync(fileName, 'utf-8');
		const lines = file.split('\n');
        const grid: number[][] = new Array<number[]>(lines.length);
        for(let i = 0; i < lines.length; i++){
            const line = lines[i];
            const parts = line.split(' ');
            grid[i] = new Array<number>(parts.length);
            for(let j = 0; j < parts.length; j++){
                const part = parts[j];
                grid[i][j] = parseInt(part, 10);
            }
        }
		return grid;
    }

    gridProduct(grid: number[][], i: number, j: number, count: number): number{
        const n = grid.length;
		let maxProduct = 1;

		let product = 1;
		if (count <= (n - i)) {
			for (let k = i; k < (i + count); k++) {
				product *= grid[j][k];
				if (product > maxProduct) {
					maxProduct = product;
				}
			}
		}

		product = 1;
		if (count <= (n - j)) {
			for (let k = j; k < (j + count); k++) {
				product *= grid[k][i];
				if (product > maxProduct) {
					maxProduct = product;
				}
			}
		}

		product = 1;
		if ((count <= (n - i)) && (count <= (n - j))) {
			for (let k = 0; k < count; k++) {
				product *= grid[j + k][i + k];
				if (product > maxProduct) {
					maxProduct = product;
				}
			}
		}

		product = 1;
		if ((count <= (n - i)) && (j > (count - 1))) {
			for (let k = 0; k < count; k++) {
				product *= grid[j - k][i + k];
				if (product > maxProduct) {
					maxProduct = product;
				}
			}
		}
		return maxProduct;
    }

    solve() {
        this.problemNumber = 11;
        let maxProduct = 1;
		let product = 1;
		const count = 4;
		const grid = this.readGrid('./p011.txt');

		const n = grid.length;
		for (let i = 0; i < n; i++) {
			for (let j = 0; j < n; j++) {
				product = this.gridProduct(grid, i, j, count);
				if (product > maxProduct) {
					maxProduct = product;
				}
			}
		}

		return maxProduct;
    }
}
